### DISCUSSION TITLE
How to scrape historical blocks / decode a block from `chain_getBlock`

### DISCUSSION THREAD

**Discussion by rossbulat** [polkadot-api/polkadot-api#879] [General]
In the case indexers or historical data aggregators need to scrape old blocks and read their data, what is the preferred approach using a typed API? 

I noticed that the [official docs](https://papi.how/typed/queries#entries-without-keys) do state that `at` can support a block hash when using `getValue`, but I did not have any joy when attempting this:

```
const historicalBlock = 21000000; // ~ 1.5 million blocks ago.
const blockHash = await client._request('chain_getBlockHash', [historicalBlock]);

// verify block number 
const verifyHistoricalBlock = await api.query.System.Number.getValue({ at: blockHash });
```

This yields the following error:
```
> NotBestBlockError: Block is not best block or finalized
...
```
Another means is to use the `chain_getBlock` RPC method directly - if this is used then what is the typedAPI function(s) needed to decode the returned scale encoded block data?

Thank you.

**Comment by josepot** [comment]
> In the case indexers or historical data aggregators need to scrape old blocks and read their data

Typically, indexers and data aggregators operate by indexing blocks as they are created. They only need to read historical data from archive nodes under specific circumstances, such as when a block was missed or when they are bootstrapping a new data set.

> What is the preferred approach using a typed API?

The typed API is specifically designed for use cases involving DApps that interact with the **current tip of the chain** or recent blocks that the DApp has directly observed. It is not intended (and arguably cannot be effectively designed) for querying arbitrary historical data. This is because the API is optimized for runtime adaptability, including handling runtime upgrades and supporting seamless integration with "**live**" chains.

If an application needs to access historical data, the most practical approach is to query that data from an indexer or data aggregator. Since accessing historical data cannot be done in a fully decentralized manner, leveraging existing indexers is the way to go.

Designing an API optimized for querying historical states or blocks from an archive node would involve significantly different considerations. Such an API would need to be tailored specifically to the requirements of accessing historical data, which is not the focus of the current typed API.

That said, the lower-level internal libraries of PAPI provide the necessary tools for accessing historical blocks and states. These tools are powerful but complex, and they should only be used by developers who fully understand the underlying constructs. It's worth noting the following caveats:
- Accessing historical state is limited to certain blocks, particularly those postdating the introduction of metadata version 14.
- Decoding ancient block bodies (extrinsics) may require custom decoders.

Currently, a typed API for querying historical data is not on the roadmap. While it’s possible this could change in the coming months, there are no concrete plans to add this capability.

### Summary:

The current typed API is not designed for reading arbitrary historical data from archive nodes. For applications needing this functionality, using an indexer or data aggregator is the recommended approach.

**Reply by rossbulat** [reply]
I would highly recommend giving an example of how to use the lower level APIs to scrape historical block data in the documentation. The current lack of support will prevent indexers from migrating fully to Polkadot API. I hope this will be placed on the roadmap in the near future.

**Reply by josepot** [reply]
> I would highly recommend giving an example of how to use the lower level APIs to scrape historical block data in the documentation.

What do you mean exactly by "scrape historical block data"? 🤔 

Do you mean decoding the header/body of an old block? or you mean querying a storage entry and/or making a runtime-calls against an arbitrary point in time? The reason why I'm asking is that if it's just for reading an old header/body, then that's fairly easy document/explain (although, it wouldn't be documented in the main polkadot-api docs). For the latter, I can already tell you that we won't document it until we decide to add support to the [`archive` group of functions](https://paritytech.github.io/json-rpc-interface-spec/api/archive.html), which is something that ATM is not even in our roadmap. 

> The current lack of support will prevent indexers from migrating fully to Polkadot API

I'm not even sure that this is completely true... Why couldn't those indexers that have already indexed the old blocks, start using PAPI for indexing the new ones? Also, many indexers use their own custom tools for indexing data, so I'm not even sure that they want to migrate to PAPI. Not to mention the fact that there are other alternatives such as subxt that they could use.

Last, but not least, I would like to remind you -once again- that we currently prioritize providing docs and libraries that empower DApp developers, and that this request falls under the category of: "other/advanced use cases".
